File: problems/77_combinations/main.cpp
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> combine(int n, int k)
    {
        vector<vector<int>> res;
        vector<int> cur;
        cur.reserve(k);

        auto dfs = [&](auto &&self, int start, int left) -> void
        {
            if (left == 0)
            { // picked k numbers
                res.push_back(cur);
                return;
            }
            // pruning: we need 'left' numbers, last possible start is n-left+1
            for (int x = start; x <= n - left + 1; ++x)
            {
                cur.push_back(x);
                self(self, x + 1, left - 1);
                cur.pop_back();
            }
        };

        dfs(dfs, 1, k);
        return res;
    }
};

string vec_to_string(const vector<int> &nums)
{
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < nums.size(); ++i)
    {
        if (i > 0)
        {
            ss << ", ";
        }
        ss << nums[i];
    }
    ss << "]";
    return ss.str();
}

string vec2d_to_string(const vector<vector<int>> &data)
{
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < data.size(); ++i)
    {
        if (i > 0)
        {
            ss << ", ";
        }
        ss << vec_to_string(data[i]);
    }
    ss << "]";
    return ss.str();
}

int main()
{
    Solution sol;

    vector<pair<int, int>> samples = {
        {4, 2},
        {1, 1}};

    vector<vector<vector<int>>> expected = {
        {{1, 2}, {1, 3}, {1, 4}, {2, 3}, {2, 4}, {3, 4}},
        {{1}}};

    for (size_t i = 0; i < samples.size(); ++i)
    {
        int n = samples[i].first;
        int k = samples[i].second;
        vector<vector<int>> result = sol.combine(n, k);

        cout << "Sample " << (i + 1) << ": n = " << n << ", k = " << k << "\n";
        cout << "Placeholder result: " << vec2d_to_string(result) << "\n";
        cout << "Expected output: " << vec2d_to_string(expected[i]) << "\n\n";
    }

    return 0;
}

File: problems/77_combinations/main.rs
struct Solution;

impl Solution {
    pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
        let mut res = Vec::new();
        let mut cur = Vec::new();

        fn dfs(start: i32, left: i32, n: i32, cur: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {
            if left == 0 {
                res.push(cur.clone());
                return;
            }
            for x in start..=n - left + 1 {
                cur.push(x);
                dfs(x + 1, left - 1, n, cur, res);
                cur.pop();
            }
        }

        dfs(1, k, n, &mut cur, &mut res);
        res
    }
}

fn main() {
    let samples = vec![
        (
            4,
            2,
            vec![
                vec![1, 2],
                vec![1, 3],
                vec![1, 4],
                vec![2, 3],
                vec![2, 4],
                vec![3, 4],
            ],
        ),
        (1, 1, vec![vec![1]]),
    ];

    for (idx, (n, k, expected)) in samples.into_iter().enumerate() {
        let result = Solution::combine(n, k);

        println!("Sample {}: n = {}, k = {}", idx + 1, n, k);
        println!("Placeholder result: {:?}", result);
        println!("Expected output: {:?}\\n", expected);
    }
}
